python/cudf/cudf/core/df_protocol.py

```diff
@@ -0,0 +1,729 @@
+        """
+        See `self.dtype` for details.
+        """
+        # Note: 'c' (complex) not handled yet (not in array spec v1).
+        #       'b', 'B' (bytes), 'S', 'a', (old-style string) 'V' (void) not handled
+        #       datetime and timedelta both map to datetime (is timedelta handled?)
+        _k = _DtypeKind
+        _np_kinds = {"i": _k.INT, "u": _k.UINT, "f": _k.FLOAT, "b": _k.BOOL,
+                     "U": _k.STRING,
+                     "M": _k.DATETIME, "m": _k.DATETIME}
+        kind = _np_kinds.get(dtype.kind, None)
+        if kind is None:
+            # Not a NumPy/CuPy dtype. Check if it's a categorical maybe
+            if isinstance(dtype, cudf.CategoricalDtype):
+                kind = _k.CATEGORICAL
+                # Codes and categorical values dtypes are different.
+                # We use codes' dtype as these are stored in the buffer. 
+                dtype = self._col.cat.codes.dtype
+            else:
+                raise ValueError(f"Data type {dtype} not supported by exchange"
+                                 "protocol")
+
+        if kind not in (_k.INT, _k.UINT, _k.FLOAT, _k.BOOL, _k.CATEGORICAL, _k.STRING):
+            raise NotImplementedError(f"Data type {dtype} not handled yet")
+
+        bitwidth = dtype.itemsize * 8
+        format_str = dtype.str
+        endianness = dtype.byteorder if not kind == _k.CATEGORICAL else '='
+        return (kind, bitwidth, format_str, endianness)
+
+    @property
+    def describe_categorical(self) -> Tuple[Any, bool, Dict[int, Any]]:
+        """
+        If the dtype is categorical, there are two options:
+
+        - There are only values in the data buffer.
+        - There is a separate dictionary-style encoding for categorical values.
+
+        Raises RuntimeError if the dtype is not categorical
+
+        Content of returned dict:
+
+            - "is_ordered" : bool, whether the ordering of dictionary indices is
+                             semantically meaningful.
+            - "is_dictionary" : bool, whether a dictionary-style mapping of
+                                categorical values to other objects exists
+            - "mapping" : dict, Python-level only (e.g. ``{int: str}``).
+                          None if not a dictionary-style categorical.
+        """
+        if not self.dtype[0] == _DtypeKind.CATEGORICAL:
+            raise TypeError("`describe_categorical only works on a column with "
+                            "categorical dtype!")
+
+        ordered = self._col.dtype.ordered
+        is_dictionary = True
+        # NOTE: this shows the children approach is better, transforming
+        # `categories` to a "mapping" dict is inefficient
+        codes = self._col.cat.codes  # ndarray, length `self.size`
+        # categories.values is ndarray of length n_categories
+        categories = self._col.cat.categories
+        mapping = {ix: val for ix, val in enumerate(categories.values_host)}
```

Comment:
This part is concerning performance-wise, since cuDF stores categories on the GPU. I'll bring this up as possibly another representation for categorical data (https://github.com/data-apis/dataframe-api/blob/main/protocol/dataframe_protocol.py#L191).

Reply:
Looks like this is already being discussed in https://github.com/data-apis/dataframe-api/issues/46.

Reply:
I opened https://github.com/data-apis/dataframe-api/pull/69 to hopefully address this.